keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,404 @@
+
+Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `UNPUBLISH_FENCE` CSI controller capability is true. If true, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+This would trigger the deletion of the `volumeAttachement` objects, which would in turn trigger `ControllerUnpublishVolume()`. The provider would need to ensure there that no volume is being used through a mount point. If unable to determine the usage, the storage provider can return an error without detaching the volume, the `external-attacher` would then retry with backoff.
+
+The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due to the shutdown.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
+
+The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.
+
+When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should forcefully cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.
+
+The Pod GC Controller should also remove the `quarantine` taint from the nodes that are already cleaned up and are safe to run pods again.
+
+### Risks and Mitigations
+
+This KEP introduces changes, and if not tested correctly could result in data corruption for users.
+To mitigate this we plan to have a high test coverage and to introduce this enhancement as an alpha
+
+## Design Details
+
+### Test Plan
+
+This feature will be tested with the following approaches:
+
+- unit tests
+- integration
+- e2e
+
+we also plan to test this with different version Skews.
+
+### Graduation Criteria
+
+This KEP will be treated as a new feature, and will be introduced with a new feature gate, `NonGracefulFailover`.
+
+This enhancement will go through the following maturity levels: alpha, beta and stable.
+
+Graduation criteria between these levels to be determined.
+
+### Upgrade / Downgrade Strategy
+
+once enabled by default, the upgrades should be smooth as the default behaviour is off by default. when downgrading/rollbacking with this behaviour enabled, there should be no issues as the control plane with the logic is still there.
```

Comment:
Also need to expand the upgrade/downgrade story for drivers that will opt-in to this.

Reply:
The current design says there is no code change for storage drivers, so we don't need it, IMO.